### PR TITLE
703 - Add util function for sorting columns of metadata TSVs

### PR DESCRIPTION
**Pull request by nozomione**
## Issue Number

Closing #703

## Purpose/Implementation Notes
This PR includes the following changes:
- Added the global sort order to the project settings
- Created a new utility function `get_sorted_field_names` 
- Used `metadata_file.write_metadata_dicts` and
   - Set  the default sort orders for columns and rows
- Cleaned up the `Project` model based on the above updates

## Types of changes

What types of changes does your code introduce?

- Refactor (addresses code organization and design mentioned in corresponding issue)
- New feature (non-breaking change which adds functionality)

## Checklist
- [x] Lint and unit tests pass locally with my changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: api/scpca_portal/common.py [hunk 1]
```diff
@@ -22,3 +22,60 @@
...
+    "min_gene_cutoff",  # REMOVED FOR SPATIAL
+    "normalization_method",  # REMOVED FOR SPATIAL
+    "demux_method",  # ONLY FOR MULTIPLEXED
+    "date_processed",
+    "workflow",
+    "workflow_version",
+    "workflow_commit",
+]
+# Rows
+METADATA_ROW_SORT_ORDER = ["scpca_project_id", "scpca_sample_id", "scpca_library_id"]
```
Comment (davidsmejia) [inline api/scpca_portal/common.py:81]: lets change this to 3 constants, `PROJECT_ID_KEY` etc.

File: api/scpca_portal/metadata_file.py [hunk 2]
```diff
@@ -92,10 +92,29 @@ def write_metadata_dicts(list_of_dicts: List[Dict], output_file_path: str, **kwa
...
+        kwargs.get("fieldnames", utils.get_keys_from_dicts(list_of_dicts))
+    )
     kwargs["delimiter"] = kwargs.get("delimiter", common.TAB)
 
+    # Set default row sort if METADATA_ROW_SORT_ORDER[0] (i.e. scpca_project_id) in list_of_dicts
+    sorted_list_of_dicts = (
+        sorted(
+            list_of_dicts,
+            key=lambda k: (
+                k[common.METADATA_ROW_SORT_ORDER[0]],
```
Comment (davidsmejia) [inline api/scpca_portal/metadata_file.py:108]: replace this with `common.PROJECT_ID_KEY` etc

File: api/scpca_portal/metadata_file.py [hunk 3]
```diff
@@ -92,10 +92,29 @@ def write_metadata_dicts(list_of_dicts: List[Dict], output_file_path: str, **kwa
...
+                k[common.METADATA_ROW_SORT_ORDER[1]],
+                k[common.METADATA_ROW_SORT_ORDER[2]],
+            ),
+        )
+        if any(common.METADATA_ROW_SORT_ORDER[0] in d for d in list_of_dicts)
```
Comment (davidsmejia) [inline api/scpca_portal/metadata_file.py:113]: TODO: Remove when library model is implemented in file creation

We do want metadata files to all have the same base "combined" metadata ie those from project and sample

File: api/scpca_portal/common.py [hunk 4]
```diff
@@ -22,3 +22,60 @@
...
+    "unfiltered_cells",
+    "filtered_cell_count",
+    "processed_cells",
+    "has_cellhash",
+    "includes_anndata",
+    "is_cell_line",
+    "is_multiplexed",
+    "is_xenograft",
+    # Project metadata
+    "scpca_project_id",
```
Comment (davidsmejia) [inline api/scpca_portal/common.py:59]: lets move this to the first item (thinking about when we have multiple projects)

File: api/scpca_portal/metadata_file.py [hunk 5]
```diff
@@ -92,10 +92,29 @@ def write_metadata_dicts(list_of_dicts: List[Dict], output_file_path: str, **kwa
     Writes a list of dictionaries to a csv-like file.
     Optional modifiers to the csv.DictWriter can be passed to function as kwargs.
     """
-    kwargs["fieldnames"] = kwargs.get("fieldnames", utils.get_keys_from_dicts(list_of_dicts))
+
+    # TODO: Only pass the default utils.get_keys_from_dicts(list_of_dicts)
+    # to utils.get_sorted_field_names once Sample.Modalities.MULTIPLEXED is removed
+    kwargs["fieldnames"] = utils.get_sorted_field_names(
+        kwargs.get("fieldnames", utils.get_keys_from_dicts(list_of_dicts))
+    )
```
Comment (davidsmejia) [inline api/scpca_portal/metadata_file.py:100]: only apply the sort to the default, it should be the responsibility of the caller to handle that

File: api/scpca_portal/models/project.py [hunk 6]
```diff
@@ -1035,7 +1038,10 @@ def write_combined_metadata_libraries(self, combined_metadata: Dict[str, List[Di
 
                 sample_metadata_path = Sample.get_output_metadata_file_path(sample_id, modality)
                 metadata_file.write_metadata_dicts(
-                    sample_libraries, sample_metadata_path, fieldnames=field_names
+                    sample_libraries,
+                    sample_metadata_path,
+                    # TODO: Remove once Sample.Modalities.MULTIPLEXED is removed
+                    fieldnames=field_names,
```
Comment (davidsmejia) [inline api/scpca_portal/models/project.py:1044]: lets remove this and have it automatically handled in this PR

**Review by davidsmejia — CHANGES_REQUESTED** [review]
Looking good just some initial feedback

File: api/scpca_portal/test/management/commands/test_load_data.py [hunk 7]
```diff
@@ -397,61 +397,63 @@ def test_multiplexed_metadata(self):
...
-            "sample_cell_count_estimate",  # with non-multiplexed
-            "sample_cell_estimates",
-            "self_reported_ethnicity_ontology_term_id",
-            "sex_ontology_term_id",
-            "tissue_ontology_term_id",
-            "transcript_type",
-            "unfiltered_cells",
-            "WHO_grade",
-        ]
+        expected_keys = utils.get_sorted_field_names(
```
Comment (davidsmejia) [inline api/scpca_portal/test/management/commands/test_load_data.py:400]: - break this up into `expected_project_keys` and `expected_sample_keys`
- remove the explicit sorting everywhere it was added as values we test against here should just be hard coded to remove hidden logic

**Review by davidsmejia — COMMENTED** [review]
added some comments about improving tests

File: api/scpca_portal/metadata_file.py [hunk 8]
```diff
@@ -92,10 +92,25 @@ def write_metadata_dicts(list_of_dicts: List[Dict], output_file_path: str, **kwa
...
+            list_of_dicts,
+            key=lambda k: (
+                k[common.METADATA_ROW_SORT_ORDER[0]],
+                k[common.METADATA_ROW_SORT_ORDER[1]],
+                k[common.METADATA_ROW_SORT_ORDER[2]],
+            ),
+        )
+        if any(common.METADATA_ROW_SORT_ORDER[0] in d for d in list_of_dicts)
+        else list_of_dicts
+    )
```
Comment (davidsmejia) [inline api/scpca_portal/metadata_file.py:111]: Could you test this, I think this makes sense as a fallback until we standardize which keys go into all of the metadata files.
```suggestion
        sorted(
            list_of_dicts,
            key=lambda k: (
                k.get(common.PROJECT_ID_KEY),
                k.get(common.SAMPLE_ID_KEY),
                k.get(common.LIBRARY_ID_KEY)
            ),
        )
    )
```

**Comment by nozomione** [comment]
I've applied your feedback and this PR is now ready for another look, thank you David!

**Review by davidsmejia — APPROVED** [review]
🎉 LGTM

**Pull request merged**